### DISCUSSION TITLE
A `POST` wihtout specifying a body

### DISCUSSION THREAD

**Discussion by Ilia-Kosenkov** [reactiveui/refit#1630] [Q&A] [answered]
Hi! Can anyone explain to me what happens if I make a `POST` request without specifying a body?
```C#
[Post("/resource/{someId}")]
Task MakePost(string someId);
```

The issue I got is that in such a scenario it seems that the first argument of the function is also sent as body (which is weird). Is this by design or is it an issue?

**Comment by EDMPhoenix** [comment]
When you make a POST request without specifying a body, the endpoint you're targeting might just treat it as an empty body or lead to an error condition depending on how that endpoint is set-up.

Regarding the situation when `someId` being sent as the body, it might be an issue specific to the framework or library you're using. Some libraries can take the first property if no body is explicitly specified. It's hard to tell without knowing the context.

Can you confirm which library/framework you are using for making the HTTP request?

**Reply by Ilia-Kosenkov** [reply]
Hi! My setup is as follows:
- I have an Azure Function with an HTTP `POST` trigger (the endpoint in question)
- I have a trivial python script that makes raw POST to this endpoint without body -- and it works as expected
- I have a trivial C# console app that does the same with `HttpClient` -- and it works as expected
- I used `Refit` to generate HTTP boilerplate for me, and when I define a `POST` method without explicitly annotating any parameter as `[Body]`, it sends the first parameter as body, if I remember correctly, serializing it as a JSON.

I own all of these apps so I can trace invocations across service boundaries, and this was what I saw. My HTTP endpoint failed because it expects either no body, or valid JSON object (so anything within `{}`), but because the first argument is a `string` (to match URL parameter), `Refit` sent `"string-content"` as a body, which failed validation (it is a valid JSON but not a valid JSON object).

Since it is the first time I am diving deep into `Refit`, I found this behavior bizarre & I wanted to confirm if it is by design before filing an issue. Unfortunately, I did not find any mention of function generator behavior in the absence of `[Body]` attribute in the docs.

I, of course, solved the problem (for now, since I was interested in just testing the endpoint without a body), by adding a `[Body] NoBody noBody` parameter, which serializes into `{}`. (Or I can actually make it nullable and default to `null` since `null` JSON is an acceptable body in my case as well).

**Reply by EDMPhoenix** [reply, marked as answer]
Refit, by design, automatically serializes the first argument of a POST method to JSON and assigns it to the request body when no parameter is explicitly set with [Body] attribute. This is why you're seeing the "string-content" in the body when Refit makes the request. 

If you don't want any body in your request, your current solution of adding a [Body] NoBody noBody parameter is good. Another way is to pass it as a null argument, like you mentioned.

The issue isn't explicitly documented because it's part of the automagic that makes Refit easy to use in most common cases.

I hope that answers your question. Don't hesitate to ask if you have any more doubts.

**Reply by Ilia-Kosenkov** [reply]
If this is not mentioned anywhere in the docs, should this be added to some readme?

**Reply by EDMPhoenix** [reply]
Yes, I think it would be a great idea to document this behavior more explicitly. While it is a design feature of Refit, it could potentially lead to confusion for new users or those not familiar with Refit's automatic [Body] assignment. 

You can propose to update the `README` or the official documentation with this detail of how Refit behaves when no [Body] attribute is specified for a POST method.

As always, contributions to open source projects like Refit help in making them better for everyone.

I hope it makes sense.
